pkg/provider/core.go

```diff
@@ -0,0 +1,388 @@
+//
+// RESPONSE PARAMETERS (driver.CreateMachineResponse)
+// ProviderID            string                   Unique identification of the VM at the cloud provider. This could be the same/different from req.MachineName.
+//                                                ProviderID typically matches with the node.Spec.ProviderID on the node object.
+//                                                Eg: gce://project-name/region/vm-ProviderID
+// NodeName              string                   Returns the name of the node-object that the VM register's with Kubernetes.
+//                                                This could be different from req.MachineName as well
+// LastKnownState        string                   (Optional) Last known state of VM during the current operation.
+//                                                Could be helpful to continue operations in future requests.
+//
+// OPTIONAL IMPLEMENTATION LOGIC
+// It is optionally expected by the safety controller to use an identification mechanisms to map the VM Created by a providerSpec.
+// These could be done using tag(s)/resource-groups etc.
+// This logic is used by safety controller to delete orphan VMs which are not backed by any machine CRD
+//
+func (p *Provider) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (*driver.CreateMachineResponse, error) {
+	// Log messages to track request
+	klog.V(2).Infof("Machine creation request has been received for %q", req.Machine.Name)
+
+	var (
+		exists       bool
+		userData     []byte
+		machine      = req.Machine
+		secret       = req.Secret
+		machineClass = req.MachineClass
+	)
+
+	// Check if incoming CR is a CR we support
+	if req.MachineClass.Provider != ProviderEquinixMetal {
+		return nil, fmt.Errorf("Requested for Provider '%s', we only support '%s'", req.MachineClass.Provider, ProviderEquinixMetal)
+	}
+
+	providerSpec, err := decodeProviderSpecAndSecret(machineClass, secret)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := p.createSVC(getApiKey(secret))
+	if svc == nil {
+		return nil, fmt.Errorf("nil Equinix Metal service returned")
+	}
+	if userData, exists = secret.Data["userData"]; !exists {
+		return nil, status.Error(codes.Internal, "userData doesn't exist")
+	}
+	// packet tags are strings only
+	createRequest := &packngo.DeviceCreateRequest{
+		Hostname:       machine.Name,
+		UserData:       string(userData),
+		Plan:           providerSpec.MachineType,
+		ProjectID:      providerSpec.ProjectID,
+		BillingCycle:   providerSpec.BillingCycle,
+		Facility:       providerSpec.Facility,
+		OS:             providerSpec.OS,
+		ProjectSSHKeys: providerSpec.SSHKeys,
+		Tags:           providerSpec.Tags,
+	}
+
+	device, _, err := svc.Devices.Create(createRequest)
+	if err != nil {
+		klog.Errorf("Could not create machine: %v", err)
+		return nil, err
+	}
+	response := &driver.CreateMachineResponse{
+		ProviderID: device.ID,
```

Comment:
I don't understand what this means

Reply:
When we create the `CreateMachineResponse` for a device created with (let us say) deviceID = abc123; we should encode this as `equinixmetal:///<facility>/<deviceID>` using the encodeMachineID() function.

You are using this `encodeMachineID()` function in the `GetMachineStatus()` driver method in line 222 to form the `GetMachineResponse`. This may not be a breaking issue, but it is good to keep the `ProviderID` field consistent across its occurrences.

Reply:
Hi @deitch ,

I think the key idea @AxiomSamarth is getting to is that the providerID field returned should match with the providerID field that appear's on the node object, refer [here](https://github.com/gardener/machine-controller-manager-provider-sampleprovider/blob/master/pkg/provider/core.go#L47-L49). Even in your previous implmentation i see similar encoding [here](https://github.com/gardener/machine-controller-manager/blob/master/pkg/driver/driver_packet.go#L168).

Reply:
Ah, that was what you meant. OK, got it and done.